### PR TITLE
Stochastic rendering option

### PR DESCRIPTION
**Pull request by lf94**
Implements a new option to adjust the chance a pixel is chosen to be rendered. This allows for faster rendering times at the cost of accuracy. All other pixels are filled in with the average of their neighbors.

### CODE REVIEW

**Comment by lf94** [comment]
@doug-moen right now there's some shader error saying "Unexpected OTHER" but searching does not bring up much. Do you have any ideas?

Also what's the correct way to access the backbuffer? Do I have to declare the uniform in the shader code in frag.cc ?

**Comment by doug-moen** [comment]
> Errors while compiling fragment shader:
> 0:260(12): preprocessor error: syntax error, unexpected OTHER

It's a preprocessor error near line 260, so it refers to this preprocessor statement:
```
#if STOCH>0
```
That is an error because STOCH is a floating point constant 0.25 and 0 is an integer constant, and you can't mix floats and integers like that in GLSL, it is a type error. So use `STOCH>0.0` instead.

**Comment by doug-moen** [comment]
The default value of STOCH is 1, so `#if STOCH > 0.0` can't be right either.
I think you must mean `#if STOCH != 1.0`.

**Comment by doug-moen** [comment]
The MacOS build is failing because that always happens when you change the GLSL code generator, due to some unit tests I wrote. So don't worry about the failed build.

**Comment by doug-moen** [comment]
Yes you would need to add a declaration for `u_backbuffer` to the generated GLSL code.

**Comment by doug-moen** [comment]
My gut feeling is that rendering at half resolution will be faster than stoch=.25 (4x faster than normal) and will look better as well. Although I haven't tested your code to see the speedup and what it looks like.

To render at half resolution, the main change is to `libcurv/viewer/viewer.cc`. Here's the first stackoverflow I found on the subject: `https://stackoverflow.com/questions/41413021/opengl-scale-single-pixel-line`. Quote:
> The only way to do this properly is to, well, do it properly. Render to an image that is actual 320x240, then draw it to the window's actual resolution.
>
> You'll have to create a texture of that size, then attach it to a framebuffer object. Bind the FBO for rendering and render to it (with the viewport set to the image's size). Then unbind the FBO, and draw that texture to the window (with the viewport set to the window's resolution).

**Comment by lf94** [comment]
Oddly enough I'm still getting that error:

[out.txt](https://github.com/curv3d/curv/files/7611982/out.txt)

**Comment by doug-moen** [comment]
Maybe the GLSL preprocessor does not support floating point arithmetic.
Try using a C++ if statement to conditionally include the STOCH code in the shader.

**Comment by doug-moen** [comment]
There is no guarantee that the backbuffer should work. I don't know how that code worked and never used it. The backbuffer references are just, from my perspective, dead code that I failed to remove when I copied and pasted from the glslViewer project. There was an initial phase where I deleted a ton of code I wasn't using, but in the end it was too much work to delete all the unused code.

Hmm. Since u_backbuffer is a uniform variable, there would need to be code in viewer.cc to initialize that uniform, and that code doesn't exist. So that's probably the issue.

**Comment by lf94** [comment]
Gotchya. I'll look into initializing it. I'm probably not going to get far...

**Comment by doug-moen** [comment]
I just found an old copy of GLSLViewer before I copied it into Curv. There's a lot of code for backbuffer support that I deleted, including 2 extra source files. Can you make this work without the backbuffer?

**Comment by lf94** [comment]
> Can you make this work without the backbuffer?

Not really because otherwise there will be a lot of black pixels, instead of pixels which were blended from the previous frame...

**Comment by lf94** [comment]
I'm trying to simply do this: get the texture data and slap it into a uniform.

From what I understand I have to use the VBO to somehow get the texture data, and then bind it, and then set a uniform...

```
    glActiveTexture(GL_TEXTURE0);
    glGenTextures(1, &mTextureID);
    glBindTexture(GL_TEXTURE_2D, mTextureID);
    glTexStorage2D(GL_TEXTURE_2D, 1, GL_RGB8, width, height);
    glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, width, height, GL_BGR, GL_UNSIGNED_BYTE, image);
```

I think the answer to this is somewhere in this block of code...

**Comment by doug-moen** [comment]
If you look at https://github.com/doug-moen/glslViewer you can find the original backbuffer code I ripped out.

In the long term I'm throwing out the OpenGL viewer code and replacing it with WebGPU. At that point, my question will be: do I want to figure out how to implement a backbuffer in WebGPU, or do I want to figure out how to implement low res rendering in WebGPU. I think that low res rendering will be simpler and will grow to be part of a more flexible infrastructure for postprocessing the frame image before displaying it in the window. The current AA code will be implemented using the same code as low res rendering (AA == render at high res then sample it down, instead of rendering at low res and sampling it up).

**Comment by lf94** [comment]
> : do I want to figure out how to implement a backbuffer in WebGPU

Absolutely not and I fully agree.

The only saving grace for me here is backbuffer is exposed through a GL_EXT. It exposes gl_LastFragData which is the pixel color at the same coordinate but from the last frame. This is enough for me to work with to get some better results. :slightly_smiling_face: 

I presume WebGPU has something similar, and if not, I'll come up with something new.

**Comment by doug-moen** [comment]
gl_LastFragData is an extension that OpenGL drivers may or may not decide to support. I don't know how widespread the support is. Curv runs on MacOS, Windows and Linux using drivers provided by Intel, AMD and NVidia. I don't want to push a change that might break on one of these OS/driver combinations. And by deciding to put the feature test for STOCH into a GLSL if statement, you are requiring every driver to support this, even if they don't turn on the STOCH feature. The glslViewer code that I pointed you at was originally tested on multiple platforms and with multiple drivers, and doesn't use this extension.

File: libcurv/frag.cc [hunk 1]
```diff
@@ -439,9 +440,22 @@ void export_frag_3d(
...
        "}\n"
 
+       "float rand(vec2 co){\n"
+       "     return fract(sin(dot(co, vec2(12.9898, 78.233))) * 43758.5453);\n"
+       "}\n"
+
        "void mainImage( out vec4 fragColour, in vec2 fragCoord )\n"
        "{\n"
-       "    vec3 col = vec3(0.0);\n"
+       "const vec4 black = vec4(0,0,0,1.0);\n"
```
Comment (doug-moen) [inline libcurv/frag.cc:449]: This variable 'black' is a new addition, and it appears to be unused.

File: libcurv/frag.cc [hunk 2]
```diff
@@ -439,9 +440,22 @@ void export_frag_3d(
...
+       "vec2 uv = (fragCoord.xy  + vec2(iTime, 0)) / iResolution.xy;\n"
```
Comment (doug-moen) [inline libcurv/frag.cc:450]: This variable 'uv' appears to be used only inside the 'if (STOCH' if statement. Please move it inside the if statement.

**Pull request merged**